api/src/main/java/org/apache/iceberg/expressions/ManifestEvaluator.java

```diff
@@ -340,6 +340,56 @@ public Boolean or(Boolean leftResult, Boolean rightResult) {
       return ROWS_MIGHT_MATCH;
     }
 
+    @Override
+    public <T> Boolean notStartsWith(BoundReference<T> ref, Literal<T> lit) {
+      int pos = Accessors.toPosition(ref.accessor());
+      PartitionFieldSummary fieldStats = stats.get(pos);
+
+      // Iceberg does not implement SQL 3-boolean logic. Therefore, for all null values, we have decided to
+      // return ROWS_MIGHT_MATCH in order to allow the query engine to further evaluate this partition, as
+      // null does not start with any non-null value.
+      if (fieldStats.containsNull() || fieldStats.lowerBound() == null) {
+        return ROWS_MIGHT_MATCH;
+      }
+
+      ByteBuffer prefixAsBytes = lit.toByteBuffer();
+
+      Comparator<ByteBuffer> comparator = Comparators.unsignedBytes();
+
+      ByteBuffer lower = fieldStats.lowerBound();
+
+      // notStartsWith will match unless all values must start with the prefix. this happens when the lower and upper
+      // bounds both start with the prefix.
+      if (lower != null) {
```

Comment:
We may want to check both lower and upper here. I'd also move the prefix handling into the block, after we know that `fieldStats.lowerBound()` and `fieldStats.upperBound()` are both non-null. I don't think that `lit.toByteBuffer()` is that expensive, but it seems like a good idea to move it just in case.